### DISCUSSION TITLE
Is it possible to use FormData and file upload?

### DISCUSSION THREAD

**Discussion by lSelectral** [ecyrbe/zodios#305] [Q&A] [answered]
This library looks promising but I didn't find any sample about, using FormData and file upload with validation.
If it possible can you provide code snippet?

**Comment by ecyrbe** [comment, marked as answer]
Hello, It's in the docs :
http://www.zodios.org/docs/client#send-multipartform-data-requests

**Reply by lSelectral** [reply]
I read the docs but it seems I missed it. Thank you so much.
But I wonder if it is possible to define FormData with all fields on the serverside?
What is the point of using this library solely on the client side?

**Reply by ecyrbe** [reply]
Zodios is both frontend and backend libray. Your FormData will all the fields is defined for both front and backend in an api definition.
Backend will receive it as a multipart-formdata and you'll still need to plug a library like multer to actually read the multipart-formdata.

**Comment by lSelectral** [comment]
@ecyrbe  Thanks a lot for your previous help.
I am using Zod server package in express app.
But I am struggling with type inference between route and authentication middleware.
Every route has its own custom **Request** and **Response** type.
But I only need type from context in this middleware. I don't need type from this specific route.
How can I write express middleware that handle Zodios types with Request context accesible?

**Reply by ecyrbe** [reply]
can you share a repository of what you are trying ?

**Reply by lSelectral** [reply]
@ecyrbe  Hello, thanks for response. 
Sorry I was busy, and couldn't reply quickly.
[Sample repository link](https://github.com/lSelectral/zodios-express-middleware-test) 

Here is what I did:
Inside: [src/config/zodios/index.ts](https://github.com/lSelectral/zodios-express-middleware-test/blob/master/src/config/zodios/index.ts) => 
1. Created a context with session object inside it
2. Created router and app from [apiSchema ](https://github.com/lSelectral/zodios-express-middleware-test/blob/master/src/config/zodios/api-schema.ts).

Then:
Created an [authentication middleware](https://github.com/lSelectral/zodios-express-middleware-test/blob/master/src/middlewares/authentication-middleware.ts). 
Authentication middleware, takes infinite amount of string parameter and return a middleware. But **BEWARE** there is **typescript error** because, I can't pass session context to middleware automatically. 

**QUESTION 1: How to pass context type to middleware?**

Then:
I imported created express app inside config to [src/app.ts](https://github.com/lSelectral/zodios-express-middleware-test/blob/master/src/app.ts)
When I try to create post route from **apiSchema**, calling middleware throw type error. 

**QUESTION 2: How can I solve this type error?**

Sorry for long post. If there is any unclarity please tell me. 
Thanks so much  ♥♥♥

**Reply by ecyrbe** [reply]
Here is how i would fix it:
https://github.com/lSelectral/zodios-express-middleware-test/pull/1

I'm using zodios express type helpers, i did not yet documented them well on zodios documentation. I concentrated more documention effort on frontend since it's the most used part atm.

Any help to improve the documentation with examples are welcomed.

**Reply by lSelectral** [reply]
@ecyrbe Thanks so much. It works like a charm.
I tried using `ZodiosRequestHandler` but couldn't figure it out how to pass to middleware object. 
I am really happy this problem solved without hacky workarounds.

Currently I am busy with multiple projects but in near future, I would like to provide examples for documentation.

THANKS A LOT AGAIN ♥♥♥

**Comment by Waqar-SE** [comment]
Hi, I was just trying to upload multiple files and i'm facing this issue where instead of file, file string representation are being uploaded, works fine when I try to upload with axios but not with zodios.

I've posted a question for it on stackoverflow as well. Will someone be able to point me to the right direction.
https://stackoverflow.com/questions/76389091/why-does-zodios-upload-a-string-instead-of-files-in-vue3
